### PR TITLE
install libcrypto locks for older libcrypto versions, until we can go…

### PR DESCRIPTION
**Pull request by JonathanHenson**
… rewrite libcrypto for s2n to use.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/s2n/s2n_tls_channel_handler.c [hunk 1]
```diff
@@ -57,6 +90,23 @@ void aws_tls_init_static_state(struct aws_allocator *alloc) {
     setenv("S2N_ENABLE_CLIENT_MODE", "1", 1);
     setenv("S2N_DONT_MLOCK", "1", 1);
     s2n_init();
+
+#if OPENSSL_VERSION_LESS_1_1
+    if (!CRYPTO_get_locking_callback()) {
+        s_libcrypto_allocator = alloc;
+        s_libcrypto_locks = aws_mem_acquire(alloc, sizeof(struct aws_mutex) * CRYPTO_num_locks());
+        assert(s_libcrypto_locks);
```
Comment (graebm) [inline source/s2n/s2n_tls_channel_handler.c]: why assert when you can exit() 🕶

File: source/s2n/s2n_tls_channel_handler.c [hunk 2]
```diff
@@ -67,6 +117,21 @@ void aws_tls_clean_up_thread_local_state(void) {
 
 void aws_tls_clean_up_static_state(void) {
     s2n_cleanup();
+
+#if OPENSSL_VERSION_LESS_1_1
+    if (CRYPTO_get_locking_callback() == s_locking_fn) {
+        CRYPTO_set_locking_callback(NULL);
```
Comment (graebm) [inline source/s2n/s2n_tls_channel_handler.c]: should we statically store whether or not we were the ones to install the locking callback?

as is, we don't set it if it was already there, but we'll remove it either way

File: source/s2n/s2n_tls_channel_handler.c (hunk 1 above)
Comment (justinboswell) [reply source/s2n/s2n_tls_channel_handler.c]: We need that hard crashing assert methinks

Comment (graebm) [reply source/s2n/s2n_tls_channel_handler.c]: aws_fatal()

File: source/s2n/s2n_tls_channel_handler.c (hunk 2 above)
Comment (JonathanHenson) [reply source/s2n/s2n_tls_channel_handler.c]: doesn't this branch check exactly that? The only way we remove it is if the locking callback is the one we would have installed.

**Pull request merged**